cube/src/main/java/zone/cogni/asquare/cube/pagination/PaginatedQuery.java

```diff
@@ -152,6 +154,39 @@ private Model getModel(RdfStoreService rdfStore, String constructQuery, long off
     return rdfStore.executeConstructQuery(batchQuery);
   }
 
+  /**
+   * Run paginated select query and handle the result row by row by the caller. No intermediate object collections are created.
+   * @param rdfStore              RDF store to use when querying
+   * @param query                 Select query that will be called with pagination. If the query already contains a limit then just this query will be executed.
+   * @param querySolutionConsumer QuerySolution consumer, will be called on each result row.
+   */
+  public void select(RdfStoreService rdfStore, String query, Consumer<QuerySolution> querySolutionConsumer) {
+    if (hasLimitAtEnd(query)) {  //is this needed? if somebody sends something with limit here it deserves to fail
+      selectBatch(rdfStore, query, querySolutionConsumer);
+      return;
+    }
+    int offset = 0;
+    while (true) {
+      String limitQuery = query + " offset " + offset + " limit " + batchSize;
+      log.info("Running query offset {} and limit {}", offset, batchSize);
+      int queryReturnedSize = selectBatch(rdfStore, limitQuery, querySolutionConsumer);
+      log.info("Query offset {} and limit {} returned {} rows", offset, batchSize, queryReturnedSize);
+      if (queryReturnedSize < batchSize) return;
+      offset += batchSize;
+    }
+  }
+
+  private int selectBatch(RdfStoreService rdfStore, String query, Consumer<QuerySolution> querySolutionConsumer) {
+    return rdfStore.executeSelectQuery(query, resultSet -> {
+      int counter = 0;
```

Comment:
maybe stream first instead of hasNext()/next()? something like:
Streams.stream(resultSet).peek(Consumer::accept).count()?

Reply:
yes but no, google guava Streams.stream is annotated as beta